halo-dev/halo
# Replace webmvc with webflux

#### What type of PR is this?

/kind cleanup
/kind optimization

#### What this PR does / why we need it:

1. Replace webmvc with webflux
2. Simplify JWT authentication procedure
3. Refine Swagger API documents
4. Remove unused controllers
5. Refactor security configuration entirely

#### Which issue(s) this PR fixes:

Fixes #2109 

#### Special notes for your reviewer:

/cc @halo-dev/sig-halo 
/area core
/milestone 2.0

Hi @halo-dev/sig-halo , you might notice that the personal access token authentication module gone, please don't feel surprised. I will refactor it after this PR is merged.

##### How to test it?

1. Authenticate with form login
    * Run Halo application
    * Open your browser and request <http://localhost:8090>
    * You will be redirected from `/` to `/login` page
    * Input username and password (for test: user/password)
    * See the default 404 page `Whitelabel Error Page`
3. Authenticate with token API
    
    * Login using cURL command
        ```bash
        curl -X POST --location "http://localhost:8090/api/auth/token" \
            -H "Accept: application/json" \
            -H "Content-Type: application/json" \
            -d "{
                \"username\": \"user\",
                \"password\": \"password\"
                }"
        ```
    * Access a resource API by using the token obtained from login
        ```bash
        curl -X GET --location "http://localhost:8090/apis/fake.halo.run/v1/posts" \
            -H "Authorization: Bearer replace_with_your_token"
        ```
    * You will get a 403 Forbidden error due to insufficient_scope

#### Does this PR introduce a user-facing change?

```release-note
None
```
